Rework MongoDB execution module due to eval deprecation

ni3mm4nd opened the pull request:
### What does this PR do?
Uses command instead of eval, adds new mongodb.version method, modifies output of mongodb.user_list and doesn't grant any roles to newly created user

### What issues does this PR fix or reference?
#39594 #33364

### Previous Behavior
mongodb.user_list returned array of arrays (each containing two tuples)
mongodb.user_create implicitly granted roles for the database (as per https://api.mongodb.com/python/current/api/pymongo/database.html#pymongo.database.Database.add_user)
mongodb.version doesn't exist

### New Behavior
mongodb.user_list returns array of dicts (keys user, roles)
mongodb.user_create grants no roles if none supplied
mongodb.version returns version of mongodb instance

### Tests written?
No

ni3mm4nd committed:
Rework MongoDB execution module due to eval deprecation

Also adds new mongodb.version method, modifies output of mongodb.user_list and doesn't grant any roles to newly created user

salt-jenkins commented:
@ni3mm4nd, thanks for your PR! By analyzing the history of the files in this pull request, we identified @rallytime, @s0undt3ch and @blueicefield to be potential reviewers.

ni3mm4nd committed:
Fix lint for MongoDB module

rallytime commented on salt/modules/mongodb.py:
```diff
@@ -223,7 +253,7 @@ def user_exists(name, user=None, password=None, host=None, port=None,
 
 
-def user_create(name, passwd, user=None, password=None, host=None, port=None,
+def user_create(name, passwd, roles=None, user=None, password=None, host=None, port=None,
```
Can you move the new `roles` kwarg to the end of the list of positional arguments? We don't want to cause an API change in case people are positional calls somewhere when calling out to this module.

ni3mm4nd replied:
It's fixed now. Sorry for that I didn't mean to break the API.

rallytime commented on salt/modules/mongodb.py:
```diff
@@ -156,6 +156,36 @@ def db_remove(name, user=None, password=None, host=None, port=None, authdb=None)
+    '''
+    conn = _connect(user, password, host, port, authdb=authdb)
+    if not conn:
+        return 'Failed to connect to mongo database'
```
Maybe it would be good to log this as an error instead of just returning the string?

ni3mm4nd replied:
Now it's fixed as you requested but I'd like to mention that almost each function in this module is returning string as i did previously. Should I modify it as well? Thanks for your comments :-)

rallytime replied:
Oh, you're right. I didn't see that the other functions were returning this way.

Now that I look more closely, I think the correct thing to do in these situations (for both this new function, as well as the others that just return a string in this module) is to raise a `CommandExecutionError`. That way it can be caught and handled correctly by other execution module functions as well as detected by the various mongodb_* state modules.

There is a new arg called `info` that you can pass into the raised exception and whatever's in that info arg is rendered using the nested outputter in the CLI return. There are some good examples of this in the [yumpkg](https://github.com/saltstack/salt/blob/5fc4c300ede98a11308cbcc958e02665de9c83bd/salt/modules/yumpkg.py#L1493-L1499) execution module, and then the the pkg state catching and handling the exception [here](https://github.com/saltstack/salt/blob/5e16de3f0064b2fed2b970c083fee7739d1fec3b/salt/states/pkg.py#L1455-L1480), for example.

rallytime reviewed (changes requested):
Thank you for updating this module @ni3mm4nd! It is very much appreciated. I had a couple of small comments/questions to address here before we can get this in.

ni3mm4nd committed:
Fix returning and logging error instead of just returning string.
Kwarg roles moved to the end of the list of positional arguments to preserve API.

ni3mm4nd committed:
Fix indentation. Replaced tabs with spaces

rallytime commented:
Thanks for pushing those changes in here @ni3mm4nd. I responded to [your question](https://github.com/saltstack/salt/pull/39595#discussion_r102885432) about the return of the other module functions.

I think making all of the functions more standardized by raising errors instead of returning strings explaining the error would be a fantastic thing to clean up in this module, and the affected states. If that is something you're willing to try, I'm happy to answer any questions about how to do that. However, I think that change could be done in a different pull request and we can still get this change in for now. 

What do you think?

ni3mm4nd commented:
Thanks for your reponse @rallytime . 

I've never done this thing with errors you mentioned before but I could give it a chance.
My plan is to contribute more in this case of MongoDB modules and states so I agree with you.
I will create separate PR for such things in the future.

Thanks once again.

rallytime commented:
Certainly! Thank you for submitting this. Let me know if you have any questions in the future.

blueicefield commented:
Seems OK for me. Thanks a lot man for your contribution.

jettero commented:
I'm surprised this didn't get a backport to 2016.11.x ...

herisanu commented:
Backport?

jettero commented:
The patch went into develop... but it's arguably a bug in 2016.11 ...
